File: src/_cpp/TowerDefenseBase/game/unit/UnitDesant.cpp
#include "UnitDesant.h"
#include "ml/common.h"
NS_CC_BEGIN





UnitDesant::UnitDesant()
: _handRadius( 60 )
, _handRadiusSector(30)
{}

UnitDesant::~UnitDesant()
{}

bool UnitDesant::init( const std::string & path, const std::string & xmlFile )
{
	return Unit::init( path, xmlFile );
}

bool UnitDesant::checkTargetByRadius( const Unit * target )const
{
	assert( target );
	Point a = _basePosition;
	Point b = target->getPosition( );
	bool byradius = checkRadiusByEllipse( a, b, getRadius( ) );
	return byradius;
}

void UnitDesant::capture_targets( const std::vector<Unit::Pointer> & targets )
{
	for( size_t i = 0; i < _targets.size(); )
	{
		auto& target = _targets[i];
		bool release = false;
		release = release || target->getMoveFinished();
		release = release || target->getCurrentHealth() <= 0;
		release = release || std::find( targets.begin(), targets.end(), target ) == targets.end();
		if( release )
			_targets.erase( _targets.begin() + i );
		else
			++i;
	}
	
	if( _targets.size() < _maxTargets )
	{
		for( auto target : targets )
		{
			auto iter = std::find( _targets.begin(), _targets.end(), target );
			if( iter == _targets.end() )
			{
				_targets.push_back( target );
			}
		}
	}
	for( auto target : _targets )
	{
		target->stop();
	}
	
	if( _targets.empty() == false )
	{
		if( current_state( ).get_name( ) != state_move &&
			isNearestTarget() == false )
		{
			Route route;
			buildRouteToTarget( route );
			getMover().setRoute( route );
			move();
		}
		else
		{
			std::vector<Unit::Pointer> t;
			t.push_back( _targets.front() );
			Unit::capture_targets( t );
		}
	}
	else
	{
		Unit::capture_targets( _targets );
		if( isNearestBase() == false )
		{
			if( current_state().get_name() != state_move )
			{
				Route route;
				buildRouteToBase( route );
				getMover().setRoute( route );
				move();
			}
		}
	}
}

bool UnitDesant::setProperty( const std::string & stringproperty, const std::string & value )
{
	if( stringproperty == "handradius" )
		_handRadius = strToFloat( value );
	else if( stringproperty == "handradiussector" )
		_handRadiusSector = strToFloat( value );
	else
		return Unit::setProperty( stringproperty, value );
	return true;
}

void UnitDesant::on_die()
{
	Unit::on_die();
	for( auto target : _targets )
	{
		target->capture_targets(std::vector<Unit::Pointer>());
		target->move();
	}
}

//void UnitDesant::on_mover( const Point & position, const Vec2 & direction )
//{
//	if( current_state().get_name() != State::state_readyfire )
//		Unit::on_mover( position, direction );
//}

void UnitDesant::buildRouteToBase( std::vector<Point> & route )
{
	route.resize( 2 );
	route[0] = getPosition();
	route[1] = _basePosition;
}

void UnitDesant::buildRouteToTarget( std::vector<Point> & route )
{
	auto target = _targets.empty() ? nullptr : _targets.front();
	if( !target )
		return;

	Point radius = target->getPosition() - getPosition();

	Point add = Point( (_handRadius - 1) * (radius.x < 0 ? 1 : -1), 0 );


	Point a = getPosition();
	Point b = target->getPosition( );
	b += add;

	route.resize( 2 );
	route[0] = a;
	route[1] = b;
}

bool UnitDesant::isNearestTarget()
{
	auto target = _targets.empty() ? nullptr : _targets.front();
	if( !target )
		return false;
	bool result;
	Vec2 radius = getPosition() - target->getPosition();
	float dist = radius.length();
	result = dist <= _handRadius;

	if( result )
	{
		float angle = getDirectionByVector( radius );
		while( angle < 0 ) angle += 360;
		
		bool r = false;
		r = r || (angle <= 0 + _handRadiusSector || angle >= 360 - _handRadiusSector);
		r = r || (angle <= 180 + _handRadiusSector || angle >= 180 - _handRadiusSector);;
		result = result && r;
	}

	return result;
}

bool UnitDesant::isNearestBase( )
{
	bool result = getPosition().getDistance( _basePosition ) < 10;
	return result;
}

NS_CC_END

File: src/_eu/game/unit/UnitDesant.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};


EU.UnitDesant = EU.Unit.extend(
{
    /** @type {Number} */ _handRadius : null ,
    /** @type {Number} */ _handRadiusSector : null,
    /** @type {Array<EU.Unit>} */ _targets : null,
    /** @type {cc.p} */ _basePosition : null,
    //CC_SYNTHESIZE_PASS_BY_REF( Point, _basePosition, BasePosition );

    setBasePosition:function(point){ this._basePosition = point; },
    getBasePosition:function(){ return this._basePosition; },

    ctor: function( path, xmlFile )
    {
        this._handRadius = 60;
        this._handRadiusSector = 30;
        this._targets = [];
        this._basePosition = cc.p(0,0) ;
        this._super(path, xmlFile);
        this.runEvent("on_stop");
    },

    checkTargetByRadius: function(   target )
    {
        EU.assert( target );
        var a = this._basePosition;
        var b = target.getPosition( );
        var byradius = EU.checkRadiusByEllipse( a, b, this._radius );
        return byradius;
    },

    capture_targets: function(targets)
    {
        for (var i = 0; i < this._targets.length; ) {
            var target = this._targets[i];
            var release = false;
            release = release || target._moveFinished;
            release = release || target.getCurrentHealth() <= 0;
            release = release || !targets || targets.indexOf(target) == -1;
            if( release )
                this._targets.splice( i , 1);
            else
                ++i;
        }

        if( targets && this._targets.length < this._maxTargets )
        {
            for (i = 0; i < targets.length; i++) {
                target = targets[i];
                if( this._targets.indexOf(target) < 0 )
                {
                    this._targets.push( target );
                }
            }
        }
        for (i = 0; i < this._targets.length; i++ ) {
            target = this._targets[i];
            target.stop();
        }

        if( this._targets.length > 0 )
        {
            if( this.current_state( ).get_name( ) != EU.MachineUnit.State.state_move && this.isNearestTarget() == false )
            {
                var route = this.buildRouteToTarget();
                this.getMover().setRoute( route );
                this.move();
            }
            else
            {
                var t = [];
                t.push( this._targets[0] );
                EU.Unit.prototype.capture_targets.call( this, t );
            }
        }
        else
        {
            EU.Unit.prototype.capture_targets.call( this, this._targets );
            if( this.isNearestBase() == false )
            {
                if( this.current_state().get_name() != EU.MachineUnit.State.state_move )
                {
                    route = this.buildRouteToBase();
                    this.getMover().setRoute( route );
                    this.move();
                }
            }
        }
    },

    setProperty_str: function( stringproperty, value )
    {
        if( stringproperty == "handradius" )
            this._handRadius = parseFloat( value );
        else if( stringproperty == "handradiussector" )
            this._handRadiusSector = parseFloat( value );
        else
            return EU.Unit.prototype.setProperty_str.call(this, stringproperty, value );
        return true;
    },

    on_die: function()
    {
        EU.Unit.prototype.on_die.call(this);
        for (var i = 0; i < this._targets.length; i++ ) {
            var target = this._targets[i];
            target.capture_targets([]);
            target.move();
        }
    },

    //on_mover: function(  var position,  var direction )
    //{
    //	if( this.current_state().get_name() != State.state_readyfire )
    //		this.on_mover( position, direction );
    //}

    buildRouteToBase: function()
    {
        var route = [];
        route[0] = this.getPosition();
        route[1] = this._basePosition;
        return route;
    },

    buildRouteToTarget: function()
    {
        var route = [];
        var target = this._targets.length == 0 ? null : this._targets[0];
        if( !target )
            return;

        var radius = cc.pSub(target.getPosition() , this.getPosition()) ;

        var add = cc.p( (this._handRadius - 1) * (radius.x < 0 ? 1 : -1), 0 );


        var a = this.getPosition();
        var b = target.getPosition( );
        cc.pAddIn(b, add);

        route[0] = a;
        route[1] = b;
        return route;
    },

    isNearestTarget: function()
    {
        var target = this._targets.length == 0 ? null : this._targets[0];
        if( !target )
            return false;
        var result;
        var radius = cc.pSub(this.getPosition() , target.getPosition()) ;
        var dist = cc.pLength(radius);
        result = dist <= this._handRadius;

        if( result )
        {
            var angle = EU.Common.getDirectionByVector( radius );
            while( angle < 0 ) angle += 360;

            var r = false;
            r = r || (angle <= this._handRadiusSector || angle >= (360 - this._handRadiusSector));
            r = r || (angle <= (180 + this._handRadiusSector) || angle >= (180 - this._handRadiusSector));
            result = result && r;
        }

        return result;
    },

    isNearestBase: function( )
    {
        var result = cc.pDistance(this.getPosition(), this._basePosition ) < 10;
        return result;
    }

});

